File: backend/internal/sales/handler/re_addressing_link.go
package handler

import (
	"context"
	"encoding/json"
	"net/http"
	"strings"

	customersservice "github.com/peter/ticket_pos/backend/internal/customers/service"
	"github.com/peter/ticket_pos/backend/internal/platform"
	"github.com/peter/ticket_pos/backend/internal/sales"
	"github.com/peter/ticket_pos/backend/internal/sales/service"
)

// The public routes a Re-addressing Link opens (#421, ADR 0058): the view
// before the button, and the click.
//
// UNDER /public, LIKE THE ASSIGNMENT LINK'S ROUTES, AND UNLIKE THEM THIS ONE
// SIGNS THE READER IN. The click is Proof of Email Ownership and the Sale is
// now theirs, so the accept returns the same sign-in a passcode does — a
// Customer Session, or the consent step when the Policy is outstanding — and
// the Storefront lands them on their tickets (#422). It is still not under
// /customer, because nobody is signed in when they arrive.

// ReAddressingSignIn is what this handler needs from customers in order to
// hand the buyer the session their click earned: the passcode's own sign-in,
// on a proven address. Declared here and satisfied by the customers service;
// unwired, the accept route refuses before it writes anything.
type ReAddressingSignIn interface {
	SignInAcceptedReAddressing(ctx context.Context, correctedEmail string) (*customersservice.SignInOutcome, error)
}

// WithReAddressingSignIn wires the sign-in an accepted Re-addressing Link
// returns. A WithX, because the customers service is built after this
// handler.
func (h *Handler) WithReAddressingSignIn(signIn ReAddressingSignIn) *Handler {
	h.reAddressingSignIn = signIn
	return h
}

// reAddressingLinkBody carries the token on the accept: in the BODY, as the
// Assignment Link's accept takes it, so the click reaches no access log and
// no Referer header.
type reAddressingLinkBody struct {
	Token string `json:"token"`
}

// ViewReAddressingLink opens a Re-addressing Link without accepting it.
//
// @Summary      View a sale re-addressing link
// @Description  Opens the Re-addressing Link a Platform Operator's Sale Re-addressing mailed to the address the buyer meant (ADR 0058), without accepting it: the Event name, the Sale Confirmation reference and the corrected address — the three facts the mail already carried — and `accepted_at` once it has been accepted. Nothing else about the purchase is disclosed before the click. The token travels in the query, as it does in the mailed link. Refused with 401 RE_ADDRESSING_LINK_INVALID when the token was tampered with, truncated, signed for another purpose or by another deployment, or names no current recording; those causes are deliberately indistinguishable. Refused with 401 RE_ADDRESSING_LINK_NO_LONGER_VALID, carrying `details.reason` of `withdrawn`, `sale_reversed` or `event_started`, when the link was genuine but its recording has ended unaccepted — told apart because its reader is the buyer.
// @Tags         public
// @Produce      json
// @Param        token  query     string  true  "The signed re-addressing link token"
// @Success      200  {object}  openapi.EnvelopeReAddressingLink
// @Failure      400  {object}  platform.Envelope
// @Failure      401  {object}  platform.Envelope
// @Failure      500  {object}  platform.Envelope
// @Router       /api/v1/public/re-addressing-link [get]
func (h *Handler) ViewReAddressingLink(w http.ResponseWriter, r *http.Request) {
	reqID := platform.RequestID(r.Context())
	token, ok := reAddressingLinkToken(w, reqID, r.URL.Query().Get("token"))
	if !ok {
		return
	}
	view, err := h.svc.ViewReAddressingLink(r.Context(), token)
	if err != nil {
		_ = platform.WriteDomainError(w, reqID, err)
		return
	}
	_ = platform.WriteSuccess(w, reqID, http.StatusOK, view)
}

// AcceptReAddressingLink accepts the Sale Re-addressing the token names and
// signs the buyer in.
//
// @Summary      Accept a sale re-addressing link
// @Description  Accepts the Sale Re-addressing the signed Re-addressing Link names (ADR 0058). The click is Proof of Email Ownership: a Customer is created or matched on the normalised corrected address and marked Verified, and in ONE transaction the Sale's Customer and snapshot email move to them, the Sale's Self-held Ticket follows if its Holder is still the wrong address, and the record is stamped accepted; a fresh Sale Confirmation is then sent to the corrected address in the Sale's own locale. First/last name, Tax ID and phone carry from the previous Customer only into a Customer nobody has named — an existing Customer's own facts win. Everything transacted stays: the Sale's reference, snapshot name, timestamps and money figures, the Payment and its snapshot, the Platform Fee, the Reversal Window (not restarted), every other Ticket's Holder and Answers, and the consent records of either Customer. **Accepting grants no consent of any kind.** **Accepting twice is idempotent**: the second click rewrites nothing, sends no second Confirmation and returns the same Sale. The response carries the Sale the buyer now owns (`ticket_sale_id`, `confirmation_ref`, `event_name`) and the sign-in outcome ON THE SAME TERMS AS A PASSCODE SIGN-IN: `session` and `session_id` when a Customer Session was minted, or `consent_required` with a pending consent token when the current Privacy Policy is outstanding for this Customer — finish it at POST /customer/auth/consent exactly as after a passcode. No cookie is set. Refused with 401 RE_ADDRESSING_LINK_INVALID for a tampered, truncated, cross-purpose or unknown token, and with 401 RE_ADDRESSING_LINK_NO_LONGER_VALID (`details.reason` of `withdrawn`, `sale_reversed` or `event_started`) when the recording has ended unaccepted; in both cases nothing is written and nobody is mailed. The Payment Provider is never called.
// @Tags         public
// @Accept       json
// @Produce      json
// @Param        body  body      reAddressingLinkBody  true  "The signed re-addressing link token"
// @Success      200  {object}  openapi.EnvelopeReAddressingAccepted
// @Failure      400  {object}  platform.Envelope
// @Failure      401  {object}  platform.Envelope
// @Failure      500  {object}  platform.Envelope
// @Router       /api/v1/public/re-addressing-link [post]
func (h *Handler) AcceptReAddressingLink(w http.ResponseWriter, r *http.Request) {
	reqID := platform.RequestID(r.Context())

	var body reAddressingLinkBody
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		_ = platform.WriteInvalidJSON(w, reqID)
		return
	}
	token, ok := reAddressingLinkToken(w, reqID, body.Token)
	if !ok {
		return
	}
	if h.reAddressingSignIn == nil {
		// Checked BEFORE the write: a Sale moved to a person who is then
		// refused a session is a worse outcome than a 500 that moved nothing.
		_ = platform.WriteDomainError(w, reqID, sales.ErrReAddressingLinkUnavailable())
		return
	}

	accepted, err := h.svc.AcceptReAddressingLink(r.Context(), token)
	if err != nil {
		_ = platform.WriteDomainError(w, reqID, err)
		return
	}
	outcome, err := h.reAddressingSignIn.SignInAcceptedReAddressing(r.Context(), accepted.CorrectedEmail)
	if err != nil {
		// The Sale has moved; the session did not come. A second click is
		// idempotent and mints it, which is what the reader will do.
		_ = platform.WriteDomainError(w, reqID, err)
		return
	}
	_ = platform.WriteSuccess(w, reqID, http.StatusOK, reAddressingAcceptedResponse(accepted, outcome))
}

func reAddressingAcceptedResponse(accepted *service.ReAddressingAccepted, outcome *customersservice.SignInOutcome) ReAddressingAcceptedResponse {
	return ReAddressingAcceptedResponse{
		TicketSaleID:    accepted.TicketSaleID,
		EventName:       accepted.EventName,
		ConfirmationRef: accepted.ConfirmationRef,
		CorrectedEmail:  accepted.CorrectedEmail,
		AcceptedAt:      accepted.AcceptedAt,
		Session:         outcome.Session,
		SessionID:       outcome.SessionID,
		ConsentRequired: outcome.ConsentRequired,
	}
}

// reAddressingLinkToken applies the one rule the token has at the handler: a
// blank one is a MALFORMED REQUEST and not an invalid link. The Storefront
// only ever sends a token it found in the address, so an empty one means the
// relay is broken rather than that somebody's link is. Everything else about
// the token is the signer's to judge.
func reAddressingLinkToken(w http.ResponseWriter, reqID, raw string) (string, bool) {
	token := strings.TrimSpace(raw)
	if token == "" {
		_ = platform.WriteValidationError(w, reqID, []platform.FieldError{
			{Field: "token", Code: platform.CodeRequired, Message: "is required"},
		})
		return "", false
	}
	return token, true
}

File: apps/storefront/lib/re-addressing-link.ts
/**
 * The Re-addressing Link's page, as rules rather than as markup (#422, parent
 * #419, ADR 0058).
 *
 * A Re-addressing Link is the Assignment Link's twin: a Platform Operator
 * recorded the address a stranded buyer meant, the platform wrote to it, and
 * pressing the link is Proof of Email Ownership (ADR 0035) — so the press mints
 * or matches a Verified Customer, moves the Sale to them, and lands them signed
 * in on it. The verb is ACCEPT and the act is a re-address; "claim" and
 * "transfer" are barred words (CONTEXT.md) and appear nowhere in this app.
 *
 * Where it differs from the Assignment Link is what may be shown BEFORE the
 * press. The Assignment page shows nothing, because showing a known Customer's
 * name would be an oracle for whether an address is registered. This page shows
 * the Event, the Sale Confirmation reference and the corrected address — the
 * three facts the mail already carried, disclosed by the API's own view
 * endpoint for exactly this purpose — and nothing else: never the buyer's name,
 * the money, the Tax ID or the Tickets. Reading the view WRITES NOTHING, which
 * is what lets a mail scanner fetch the page and change nothing.
 *
 * Pure and dependency-free — no React, no i18n runtime — so the rules are
 * directly unit-testable. It returns SHAPES and never a sentence.
 */

import { customerAreaSaleHref } from "./destination.ts";
import { signInConsentPath } from "./google-signin.ts";

/**
 * What GET /public/re-addressing-link shows before the press: the three facts
 * the mail carried, and whether the link has already been accepted.
 *
 * `accepted_at` is the one fact the mail could not carry. When it is set the
 * page says the purchase is already theirs and STILL offers the button: a mail
 * opened twice, or a button pressed twice, must land the reader on the Sale
 * rather than on an error, and the accept is idempotent on the API's side so
 * that pressing it again rewrites nothing.
 */
export type ReAddressingLinkView = {
  event_name: string;
  confirmation_ref: string;
  corrected_email: string;
  accepted_at: string | null;
};

/**
 * What the BFF answers after a press: the Sale the reader now owns, and whether
 * the Privacy Policy still has to be accepted before a session exists.
 *
 * The session itself never reaches the browser. The BFF holds it in the httpOnly
 * cookie exactly as the passcode verify route does, and when consent is
 * outstanding it parks the pending-consent token in the cookie the Google
 * callback uses, so the sign-in page's own consent step finishes the job.
 */
export type ReAddressingAccepted = {
  ticket_sale_id: string;
  consent_required: boolean;
};

/**
 * The token out of the address, or "" when the address does not carry one.
 *
 * A REPEATED parameter is not a link this platform ever wrote, and taking the
 * first would be guessing at which one somebody meant; so is a token that is
 * only whitespace. The API trims too, so trimming here is not a second
 * definition of what a token is — it is what lets the page tell "no token" from
 * "a token the API refused" before making a request.
 */
export function reAddressingLinkToken(raw: string | string[] | undefined): string {
  return typeof raw === "string" ? raw.trim() : "";
}

/**
 * The `reAddressingLink` catalog key for each way the link can be refused.
 *
 * Unlike the Assignment Link, this page TELLS THE REASONS APART. The
 * Assignment page's reader does not know who bought the ticket, so "your friend
 * gave it to someone else" is a fact about a stranger's decision it must not
 * disclose. This page's reader is the BUYER: it was their purchase, their money
 * and their mistyped address, and the API carries the reason in
 * `details.reason` for that reason alone. A withdrawn re-addressing, a reversed
 * Sale and a started Event are each said plainly.
 */
export const RE_ADDRESSING_LINK_FAILURE_KEYS = {
  // Tampered, truncated, signed for another purpose, or naming a recording
  // that a later one replaced. Deliberately indistinguishable on the API's
  // side; one sentence here.
  RE_ADDRESSING_LINK_INVALID: "invalid",
  // No signing key configured: a deployment fault, not the reader's, so it
  // says "try again later" rather than blaming a link they cannot replace.
  RE_ADDRESSING_LINK_UNAVAILABLE: "unavailable",
} as const;

/**
 * The reasons a genuine link no longer works, in the API's own spelling
 * (sales.DeriveReAddressingState) and the catalog key each one reads as.
 */
export const RE_ADDRESSING_LINK_ENDED_KEYS = {
  withdrawn: "withdrawn",
  sale_reversed: "saleReversed",
  event_started: "eventStarted",
} as const;

export type ReAddressingLinkFailure =
  | (typeof RE_ADDRESSING_LINK_FAILURE_KEYS)[keyof typeof RE_ADDRESSING_LINK_FAILURE_KEYS]
  | (typeof RE_ADDRESSING_LINK_ENDED_KEYS)[keyof typeof RE_ADDRESSING_LINK_ENDED_KEYS];

/**
 * Which copy a refused link gets, from the API's error code and details.
 *
 * `RE_ADDRESSING_LINK_NO_LONGER_VALID` is the only code that carries a reason,
 * and a reason this app has not heard of falls back to "invalid" — the honest
 * floor, since the link does not work and the page has nothing truer to say. So
 * does a code it has not heard of at all.
 */
export function reAddressingLinkFailure(
  code: string | undefined,
  details?: unknown,
): ReAddressingLinkFailure {
  if (code === "RE_ADDRESSING_LINK_NO_LONGER_VALID") {
    const reason =
      typeof details === "object" && details !== null
        ? (details as { reason?: unknown }).reason
        : undefined;
    if (typeof reason === "string" && reason in RE_ADDRESSING_LINK_ENDED_KEYS) {
      return RE_ADDRESSING_LINK_ENDED_KEYS[reason as keyof typeof RE_ADDRESSING_LINK_ENDED_KEYS];
    }
    return "invalid";
  }
  if (code && code in RE_ADDRESSING_LINK_FAILURE_KEYS) {
    return RE_ADDRESSING_LINK_FAILURE_KEYS[code as keyof typeof RE_ADDRESSING_LINK_FAILURE_KEYS];
  }
  return "invalid";
}

/**
 * What the page renders, decided once from the address and the view read.
 *
 * `incomplete` is a link that arrived without its token — truncated by a mail
 * client, or copied by hand — and is told apart from a refusal because there is
 * nothing to send the API. `ready` and `accepted` both offer the button; the
 * second says the purchase is already theirs first. Every `failed` state is a
 * plain "this link no longer works" with the reason above and a way back to
 * the Storefront, since this reader has an ordinary account to go to rather
 * than a stranger's Ticket to wonder about.
 */
export type ReAddressingLinkState =
  | { kind: "incomplete" }
  | { kind: "ready"; view: ReAddressingLinkView }
  | { kind: "accepted"; view: ReAddressingLinkView }
  | { kind: "failed"; failure: ReAddressingLinkFailure };

/** The view read's outcome, as the page hands it over: the view, or the API's refusal. */
export type ReAddressingLinkRead =
  | { status: "ok"; view: ReAddressingLinkView }
  | { status: "error"; code: string | undefined; details?: unknown };

export function reAddressingLinkState(
  token: string,
  read: ReAddressingLinkRead | null,
): ReAddressingLinkState {
  if (token === "") {
    return { kind: "incomplete" };
  }
  if (!read || read.status === "error") {
    return { kind: "failed", failure: reAddressingLinkFailure(read?.code, read?.details) };
  }
  return read.view.accepted_at
    ? { kind: "accepted", view: read.view }
    : { kind: "ready", view: read.view };
}

/**
 * Where the browser goes after a successful press.
 *
 * With a session set, straight to the Sale in the Customer Area — the same
 * anchor the Confirmation Link and the checkout success page use, so the reader
 * lands on the purchase and not merely on the list. With consent outstanding,
 * to the sign-in page's consent step with that same Sale as the destination,
 * exactly where the Google callback sends a first-time Customer: one consent
 * step, one submission endpoint, and the sign-in page pushes them on to the
 * Sale when it is done.
 */
export function reAddressingLinkNext(accepted: ReAddressingAccepted): string {
  const sale = customerAreaSaleHref(accepted.ticket_sale_id);
  return accepted.consent_required ? signInConsentPath(sale) : sale;
}
